File: 2021/day1/day1.cpp
//
//  main.cpp
//  day1
//
//  Created by 박경선 on 2021/12/01.
//

#include <iostream>
#include <fstream>
#include <vector>
#include <climits>
#include <cstring>
#include <unistd.h>

using namespace std;

void part1(vector<int> input) {
    int answer = 0;
    
     for (int i = 1; i < input.size(); i++) {
        if (input[i] > input[i-1]) answer++;
     }
     
     cout << ":: part1 answer is " << answer << endl;
}


void part2(vector<int> input) {
    int answer = 0;
    int prev = INT_MAX;

    for (int i = 0; i < input.size()-2; i++) {
        if (input[i] + input[i+1] + input[i+2] > prev) answer++;
        prev = input[i] + input[i+1] + input[i+2];
    }

    cout << ":: part2 answer is " << answer << endl;
}

string getFilename(int argc, char * argv[]) {
    for (int i = 1; i < argc; i++) {  
        if (i + 1 != argc) { 
            if (strcmp(argv[i], "-f") == 0) {                 
                return argv[i + 1];
            }
        }
    }
    return "";
}

int main(int argc, char * argv[]) {
    vector<int> input;
    string filePath = "./data/input.txt";

    if (fopen(argv[2], "r") != NULL) {
        filePath = getFilename(argc, argv);
    }

    ifstream file(filePath);
    if(file.is_open()){
        string line;
        while(getline(file, line)){
            input.push_back(stoi(line));
        }
        file.close();
    }

    part1(input);
    part2(input);
   
    return 0;
}

File: 2021/day1/day1.go
package main

import (
	"bufio"
	"flag"
	"log"
	"math"
	"os"
	"strconv"
)

var filepath = flag.String("f", "./data/input.txt", "a string of file path")

func check(e error) {
	if e != nil {
		panic(e)
	}
}

func part1(input []int) {
    answer := 0
	for i := 1; i < len(input); i++ {
        if (input[i] > input[i-1]) {answer++;}
	}

	println(":: part1 answer is ", answer)
}

func part2(input []int) {
    answer := 0
	prev := math.MaxInt

	for i := 0; i < len(input)-2; i++ {
        if (input[i] + input[i+1] + input[i+2] > prev) { answer++ }
		prev = input[i] + input[i+1] + input[i+2]
     }

	println(":: part2 answer is ", answer)
}

func init() {
	flag.Parse()
}

func main() {
	var input = []int{}

	f, err := os.Open(*filepath)
	check(err)
	defer f.Close()
	reader := bufio.NewScanner(f)

	for reader.Scan() {
		sv, err := strconv.Atoi(reader.Text())
		check(err)
		input = append(input, sv)
    }

    if err := reader.Err(); err != nil {
        log.Fatalf("Error while reading file: %s", err)
    }

	part1(input)
	part2(input)
}